### PR TITLE
Extract article metadata

### PR DESCRIPTION
**Pull request by jfmontanaro**
Adds metadata extraction. 

I decided to split the page title (contents of the `<title>` element) apart from the "article title" (content of a matching meta tag, or of a h1/h2 if it's the only one on the page) because these are often different. For example, a lot of sites append or prepend the site name to the HTML `<title>`, and I don't think we want that showing up in the page header. 

If it finds one but not the other, it will return the same value for both.

@mre You mentioned you'd like the title to be passed through unmodified, rather than being altered during extraction. Unfortunately it seems that calling `NodeRef::text_contents()` automatically unescapes HTML entities, e.g. `&quot;` becomes `"`. Do you want me to see if I can find a way around that, or were you thinking of other modifications?

### CODE REVIEW

File: src/metadata.rs
Comment (mre) [inline src/metadata.rs:54]: Can this panic? If so you can convert it to an option:

```suggestion
    let mut h1s = root.select("h1").ok();
```

Same for the h2 down there.

Comment (mre) [reply src/metadata.rs:54]: Ah, you mention that later: "unwrap is safe here because select() only errors when you give it an invalid CSS selector". Nevermind then. Still a nice trick that could come in handy in the future. 😉

**Review by mre — APPROVED** [review]
Very well done. No objections from my side. 👍

**Comment by mre** [comment]
> @mre You mentioned you'd like the title to be passed through unmodified, rather than being altered during extraction. Unfortunately it seems that calling NodeRef::text_contents() automatically unescapes HTML entities, e.g. &quot; becomes ". Do you want me to see if I can find a way around that, or were you thinking of other modifications?

I think what you did in this PR makes a lot of sense and we can keep it that way. My main gripe with the previous approach was that the algorithm was too opinionated and kept modifying the title heavily (adding the website name to the title, messing with the spacing etc.). Judging from your test cases it looks like you struck a nice balance between cleaning up the title without mangling it. So all good from my end.

I also liked that you added additional metadata as well.

File: src/metadata.rs
Comment (jfmontanaro) [reply src/metadata.rs:54]: Sorry, yeah originally that comment appeared the first time I called `unwrap()` in that situation, but then I refactored and forgot to move the comment. :smile:

Comment (mre) [reply src/metadata.rs:54]: Haha all good. Looking forward to seeing this live. Feel free to merge and publish. I'll do the shuttle part again.

**Comment by jfmontanaro** [comment]
Ok, going to go ahead and merge this. I was holding off because I discovered some possibly-more-refined byline parsing already exists in the codebase, so I was wondering if I could include that as well somehow. However it appears to be turned off for some reason, and I don't understand it well enough to know why at this point, so I'll leave it for now.

**Comment by mre** [comment]
Thanks, will you make the changes on the `readable` repo? I will go ahead and add continuous deployment so it will go live once we merge over there. (Although it will be slow.)

**Comment by jfmontanaro** [comment]
Done - you weren't kidding about it being slow, though. Do you mind if I take a look into caching the build artifacts at some point? I know GH Actions has a built-in cache solution, but I think it's pretty aggressive about deleting your data if you don't run a build for a while. I've heard good things about [sccache](https://github.com/mozilla/sccache#storage-options), I might look into that, possibly with Backblaze as the storage backend since they give you 10GB for free.

**Comment by mre** [comment]
Go for it. If you find a way to make it faster I'm sure the shuttle team would be interested as well.

**Pull request merged**
